pkg/controller/machineset/machineset_controller_test.go

```diff
@@ -17,196 +17,134 @@ limitations under the License.
 					},
 				},
 			},
-			expectedRequest: reconcile.Request{NamespacedName: types.NamespacedName{Name: "foo", Namespace: "default"}},
-			// Verify machines are created and recreated after deletion
-			verifyFnc: func() {
-				machines := &machinev1beta1.MachineList{}
-
-				// TODO(joshuarubin) there seems to be a race here. If expectInt sleeps
-				// briefly, even 10ms, the number of replicas is 4 and not 2 as expected
-				expectInt(t, int(replicas), func(ctx context.Context) int {
-					if err := c.List(ctx, machines); err != nil {
-						return -1
-					}
-					return len(machines.Items)
-				})
-
-				// Delete a Machine and expect Reconcile to be called to replace it.
-				m := machines.Items[0]
-				if err := c.Delete(context.TODO(), &m); err != nil {
-					t.Errorf("error deleting machine: %v", err)
-				}
-				select {
-				case recv := <-requests:
-					if recv != expectedRequest {
-						t.Error("received request does not match expected request")
-					}
-				case <-time.After(timeout):
-					t.Error("timed out waiting for request")
-				}
-
-				// TODO (robertbailey): Figure out why the control loop isn't working as expected.
-				/*
-					g.Eventually(func() int {
-						if err := c.List(context.TODO(), machines); err != nil {
-							return -1
-						}
-						return len(machines.Items)
-					}, timeout).Should(gomega.BeEquivalentTo(replicas))
-				*/
-			},
-		},
-	}
-
-	for _, tc := range testCases {
-		t.Logf("Running %q testcase", tc.name)
-		func() {
-			if err := c.Create(context.TODO(), tc.instance); err != nil {
-				t.Errorf("error creating instance: %v", err)
-			}
+		}
 
-			defer func() {
-				err := c.Delete(context.TODO(), tc.instance)
-				if err != nil {
-					t.Logf("error deleting %v: %v", tc.instance, err)
-				}
+		By("Creating the MachineSet")
+		Expect(k8sClient.Create(ctx, instance)).To(Succeed())
 
-				select {
-				case recv := <-requests:
-					if recv != tc.expectedRequest {
-						t.Error("received request does not match expected request")
-					}
+		machines := &machinev1.MachineList{}
 
-					t.Logf("Last reconcile %v", <-requests)
-				case <-time.After(timeout):
-					t.Error("timed out waiting for request")
-				}
-			}()
+		By("Verifying that we have 2 replicas")
+		Eventually(func() (int, error) {
+			if err := k8sClient.List(ctx, machines, client.InNamespace(namespace.Name)); err != nil {
+				return 0, err
+			}
+			return len(machines.Items), nil
+		}, timeout).Should(BeEquivalentTo(replicas))
+
+		By("Trying to delete 1 machine and check the MachineSet scales back up")
+		machineToBeDeleted := machines.Items[0]
+		Expect(k8sClient.Delete(ctx, &machineToBeDeleted)).To(Succeed())
+
+		// Verify that the Machine has been deleted.
+		Eventually(func() bool {
+			key := client.ObjectKey{Name: machineToBeDeleted.Name, Namespace: machineToBeDeleted.Namespace}
+			if err := k8sClient.Get(ctx, key, &machineToBeDeleted); apierrors.IsNotFound(err) || !machineToBeDeleted.DeletionTimestamp.IsZero() {
```

Comment:
Is there anything adding a finalizer to the Machines? Do you expect them not to be deleted straight away?, I'd be tempted to remove the second check and see if this still passes?

Reply:
You are right, the check can be removed